### PR TITLE
Move root garminbackup.py script into garminexport and break it into two functions

### PR DESCRIPTION
**Pull request by khrapovs**
This change has the same purpose as https://github.com/petergardfjall/garminexport/pull/43. But here instead of passing a large number of arguments like `username`, `password`, etc., one has to pass a single object `args` with corresponding attributes used in the original script. No regressions are introduced, the root script is still there. The benefit is that one can use this library in another python script like so:
```
from garminexport.garminbackup import garminbackup
args = MyArgs(username=garmin_login, password=garmin_password,
              file_format=['json_summary'], backup_dir=os.path.join(path_garmin_files, 'json'))
garminbackup(args=args)
```
Thanks to @tgrecojr for seeing that I am not alone in need for such a change.

### CODE REVIEW

File: garminexport/garminbackup.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    "WARNING": logging.WARNING,
+    "ERROR": logging.ERROR
+}
+"""Command-line (string-based) log-level mapping to logging module levels."""
+
+DEFAULT_MAX_RETRIES = 7
+"""The default maximum number of retries to make when fetching a single activity."""
+
+
+def parse_args():
```
Comment (petergardfjall) [inline garminexport/garminbackup.py]: I would prefer to have this in a separate file: `garminexport/cli.py`.
Also, add a brief description of the function.

File: garminexport/garminbackup.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
+#! /usr/bin/env python
+"""Performs (incremental) backups of activities for a given Garmin Connect
```
Comment (petergardfjall) [inline garminexport/garminbackup.py]: This comment belongs in the function head of `garminbackup()`

File: garminexport/garminbackup.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+from garminexport.backup import export_formats
+from garminexport.retryer import (
+    Retryer, ExponentialBackoffDelayStrategy, MaxRetriesStopStrategy)
+import logging
+import os
+import re
+import sys
+import traceback
+
+logging.basicConfig(
```
Comment (petergardfjall) [inline garminexport/garminbackup.py]: If this is to be used as a library, I'd like the library client to be able to determine logging behavior. Therefore, move all log-related setup out to the "main" script.

File: garminexport/garminbackup.py (hunk 1 above)
Comment (petergardfjall) [reply garminexport/garminbackup.py]: Actually, also mention that the returned dict can be used directly with the `garminbackup()` function.

File: garminexport/garminbackup.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        help="Ignore errors and keep going. Default: FALSE")
+    parser.add_argument(
+        "--max-retries", metavar="NUM", default=DEFAULT_MAX_RETRIES,
+        type=int, help="The maximum number of retries to make on failed attempts to fetch an activity. Exponential backoff will be used, meaning that the delay between successive attempts will double with every retry, starting at one second. DEFAULT: %d" % DEFAULT_MAX_RETRIES)
+
+    return parser.parse_args()
+
+
+def garminbackup(args=None):
+
```
Comment (petergardfjall) [inline garminexport/garminbackup.py]: This PR aims to make `garminbackup` a publically exposed API. IMHO, a "anonymous" dict doesn't cut it for an API. We need to be explicit about the arguments. Add one key-value argument for every parameter that can be passed as an arg and document the function API.

File: garminexport/garminbackup.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    if not args.log_level in LOG_LEVELS:
+        raise ValueError("Illegal log-level: {}".format(args.log_level))
+
+    # if no --format was specified, all formats are to be backed up
+    args.format = args.format if args.format else export_formats
+    log.info("backing up formats: %s", ", ".join(args.format))
+
+    logging.root.setLevel(LOG_LEVELS[args.log_level])
```
Comment (petergardfjall) [inline garminexport/garminbackup.py]: I don't think the user of the library wants this function to reset the root logger's level as a side-effect. This belongs in "main".

File: garminexport/garminbackup.py (hunk 4 above)
Comment (petergardfjall) [reply garminexport/garminbackup.py]: Also, describe what the function does and where/how it stores downloaded content.

Comment (petergardfjall) [reply garminexport/garminbackup.py]: Also, I wonder if `incremental_backup()` would be a better name.

File: garminexport/garminbackup.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                    garminexport.backup.download(
+                        client, activity, retryer, args.backup_dir,
+                        args.format)
+                except Exception as e:
+                    log.error(u"failed with exception: %s", e)
+                    if not args.ignore_errors:
+                        raise
+    except Exception as e:
+        exc_type, exc_value, exc_traceback = sys.exc_info()
+        log.error(u"failed with exception: %s", str(e))
```
Comment (petergardfjall) [inline garminexport/garminbackup.py]: Any errors should probably be propagated all the way to the initial caller. This exception handling belongs in "main".

**Comment by khrapovs** [comment]
Nice! Thanks for a very detailed review! If you are quick to respond, I am ready to invest much more in this package.

File: garminexport/garminbackup.py (hunk 1 above)
Comment (khrapovs) [reply garminexport/garminbackup.py]: done. cli.py is there

File: garminexport/garminbackup.py (hunk 2 above)
Comment (khrapovs) [reply garminexport/garminbackup.py]: moved. but left the original comment block with some cosmetic changes

File: garminexport/garminbackup.py (hunk 3 above)
Comment (khrapovs) [reply garminexport/garminbackup.py]: dealing with logging is not my strongest. I tried to rearrange logging related lines to the best of my understanding. for example, there is a new `garminexport/logging_config.py` that stores `LOG_LEVELS` dictionary. would appreciate some more specific advice if this change is not what you expect.

File: garminexport/garminbackup.py (hunk 4 above)
Comment (khrapovs) [reply garminexport/garminbackup.py]: changed the name of the function. parametrized it with the same arguments as requested by arg parser

File: garminexport/garminbackup.py (hunk 5 above)
Comment (khrapovs) [reply garminexport/garminbackup.py]: again, simply moved it to the root script. not sure that it is correct though

File: garminexport/garminbackup.py (hunk 6 above)
Comment (khrapovs) [reply garminexport/garminbackup.py]: moved to the root script. btw, what purpose serves this line: ```exc_type, exc_value, exc_traceback = sys.exc_info()```? returned objects do not seem to be used anywhere...

**Comment by petergardfjall** [comment]
> Nice! Thanks for a very detailed review! If you are quick to respond, I am ready to invest much more in this package.

Thanks for the changes! I apologize for not having looked at this sooner. I've been a bit busy.
I'll try to find some time to look at it more closely in the coming days.

File: garminbackup.py [hunk 7]
```diff
@@ -1,124 +1,32 @@
...
-            log.info("%s contains %d backed up activities",
-                args.backup_dir, len(backed_up))
-
-            log.info("activities that aren't backed up: %d",
-                     len(missing_activities))
+        incremental_backup(username=args.username,
+                           password=args.password,
+                           backup_dir=args.backup_dir,
+                           format=args.format,
+                           log_level=args.log_level,
```
Comment (petergardfjall) [inline garminbackup.py]: I would say we skip the `log_level` argument here. Instead `incremental_backup` will simply use a logger named after its module (`garminexport.incremental_backup`) and it is up to the client to modify the log level for that logger if need be:
```
   logging.getLogger('garminexport.incremental_backup').setLevel(logging.ERROR)
```
or similar.

File: garminexport/incremental_backup.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+from garminexport.retryer import Retryer, ExponentialBackoffDelayStrategy, MaxRetriesStopStrategy
+
+log = logging.getLogger(__name__)
+
+
+def incremental_backup(username: str,
+                       password: str = None,
+                       backup_dir: str = os.path.join(".", "activities"),
+                       format: str = 'ALL',
+                       log_level: str = 'INFO',
```
Comment (petergardfjall) [inline garminexport/incremental_backup.py]: Let's remove this argument.

File: garminexport/incremental_backup.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    :param max_retries: The maximum number of retries to make on failed attempts to fetch an activity.
+    Exponential backoff will be used, meaning that the delay between successive attempts
+    will double with every retry, starting at one second. Default: 7.
+
+    The activities are stored in a local directory on the user's computer.
+    The backups are incremental, meaning that only activities that aren't already
+    stored in the backup directory will be downloaded.
+    """
+    if log_level not in LOG_LEVELS:
+        raise ValueError("Illegal log-level: {}".format(log_level))
```
Comment (petergardfjall) [inline garminexport/incremental_backup.py]: We can drop this check without the `log_level` argument.

File: garminexport/garminbackup.py (hunk 6 above)
Comment (petergardfjall) [reply garminexport/garminbackup.py]: Must be a leftover from an older version of the code. Feel free to clean up! :)

Comment (khrapovs) [reply garminexport/garminbackup.py]: removed the line

File: garminbackup.py (hunk 7 above)
Comment (khrapovs) [reply garminbackup.py]: `log_level` argument is gone. I did not notice that it is not used after refactoring

File: garminexport/incremental_backup.py (hunk 8 above)
Comment (khrapovs) [reply garminexport/incremental_backup.py]: removed

File: garminexport/incremental_backup.py (hunk 9 above)
Comment (khrapovs) [reply garminexport/incremental_backup.py]: removed

**Comment by petergardfjall** [comment]
This was a substantial improvement of the code base, taking it in a nice library direction. 
Thank you @khrapovs!

**Pull request merged**